File: solutions/python3/medium/binary-tree-library-book-collection-tracker/solution-approach1.py
# ──────────────────────────────────────────────────
# Link        https://www.hackerrank.com/contests/ti27110-sri-sai-ram/challenges/binary-tree-library-book-collection-tracker/problem?isFullScreen=true
# Problem     Binary Tree - Library Book Collection Tracker
# Difficulty  Medium
# Subdomain   N/A
# Platform    HackerRank
# Language    python3
# Status      Accepted
# Submitted   2026-08-03, 09:11 a.m.
# ──────────────────────────────────────────────────

import sys
from collections import deque

class Node:
    """Represents a node (book) in the Binary Tree."""
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BinaryTree:
    """Class to manage Binary Tree operations."""
    def __init__(self):
        self.root = None

    def insert_level_order(self, book_ids):
        """Inserts book IDs into the binary tree level by level."""
        if not book_ids:
            return

        self.root = Node(book_ids[0])
        queue = deque([self.root])
        i = 1

        while i < len(book_ids):
            current = queue.popleft()

            # Insert left child
            if i < len(book_ids):
                current.left = Node(book_ids[i])
                queue.append(current.left)
                i += 1

            # Insert right child
            if i < len(book_ids):
                current.right = Node(book_ids[i])
                queue.append(current.right)
                i += 1

    def count_nodes(self, node):
        """Recursively counts the total number of nodes in the binary tree."""
        if node is None:
            return 0
        return 1 + self.count_nodes(node.left) + self.count_nodes(node.right)


def main():
    # Read all inputs from standard input
    input_data = sys.stdin.read().split()
    if not input_data:
        return

    n = int(input_data[0])
    book_ids = [int(x) for x in input_data[1:n+1]]

    tree = BinaryTree()
    tree.insert_level_order(book_ids)

    # Count nodes using the recursive method
    total_books = tree.count_nodes(tree.root)
    print(total_books)

if __name__ == "__main__":
    main()

File: solutions/java15/medium/binary-search-tree-customer-age-based-filtering/solution-approach1.java
// ──────────────────────────────────────────────────
// Link        https://www.hackerrank.com/contests/ti27110-sri-sai-ram/challenges/binary-search-tree-customer-age-based-filtering/problem?isFullScreen=true
// Problem     Binary Search Tree - Customer Age-Based Filtering
// Difficulty  Medium
// Subdomain   N/A
// Platform    HackerRank
// Language    java15
// Status      Accepted
// Submitted   2026-08-03, 10:42 p.m.
// ──────────────────────────────────────────────────

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
        this.val = val;
    }
}

public class Solution {

    private static TreeNode insertIntoBST(TreeNode root, int val) {
        if (root == null) {
            return new TreeNode(val);
        }
        if (val < root.val) {
            root.left = insertIntoBST(root.left, val);
        } else if (val > root.val) {
            root.right = insertIntoBST(root.right, val);
        }
        return root;
    }

    private static TreeNode trimBST(TreeNode root, int low, int high) {
        if (root == null) {
            return null;
        }

        if (root.val < low) {
            return trimBST(root.right, low, high);
        }
        if (root.val > high) {
            return trimBST(root.left, low, high);
        }

        root.left = trimBST(root.left, low, high);
        root.right = trimBST(root.right, low, high);

        return root;
    }

    private static void preorderTraversal(TreeNode root, List<Integer> result) {
        if (root != null) {
            result.add(root.val);
            preorderTraversal(root.left, result);
            preorderTraversal(root.right, result);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        if (!scanner.hasNextInt()) {
            return;
        }

        int n = scanner.nextInt();
        TreeNode root = null;

        for (int i = 0; i < n; i++) {
            int age = scanner.nextInt();
            root = insertIntoBST(root, age);
        }

        if (!scanner.hasNextInt()) {
            return;
        }
        int low = scanner.nextInt();

        if (!scanner.hasNextInt()) {
            return;
        }
        int high = scanner.nextInt();

        root = trimBST(root, low, high);

        List<Integer> result = new ArrayList<>();
        preorderTraversal(root, result);

        for (int i = 0; i < result.size(); i++) {
            System.out.print(result.get(i) + (i == result.size() - 1 ? "" : " "));
        }
        System.out.println();

        scanner.close();
    }
}
